runtime/expr/function/unflatten.go

```diff
@@ -114,22 +135,44 @@ type record struct {
 	records []*record
 }
 
-func (r *record) addPath(c *recordCache, p []string) {
+func (r *record) addPath(c *recordCache, p []string) (removed int) {
 	if len(p) == 0 {
-		return
+		return 0
 	}
-	if len(r.columns) == 0 || r.columns[len(r.columns)-1].Name != p[0] {
+	at := len(r.columns) - 1
+	if len(r.columns) == 0 || r.columns[at].Name != p[0] {
 		r.columns = append(r.columns, zed.NewColumn(p[0], nil))
 		var rec *record
 		if len(p) > 1 {
 			rec = c.new()
 		}
 		r.records = append(r.records, rec)
+	} else if len(p) == 1 || r.records[at] == nil {
+		// If this isn't a new column and we're either at a leaf or the
+		// previously value was a leaf, we're stacking on a previously created
+		// record and need to signal that values have been removed.
+		removed = countleaves(r.records[at])
+		if len(p) > 1 {
+			r.records[at] = c.new()
+		} else {
+			r.records[at] = nil
+		}
 	}
-	r.records[len(r.records)-1].addPath(c, p[1:])
+	removed += r.records[len(r.records)-1].addPath(c, p[1:])
+	return removed
 }
 
-func (r *record) build(zctx *zed.Context, b *zcode.Builder, next func() (zed.Type, zcode.Bytes)) zed.Type {
+func countleaves(rec *record) (count int) {
```

Comment:
Nits: Make this a method, call it `countLeaves` for consistency with all the other mixedCaps names in this file, and don't use a named result parameter just to save a line.